### DISCUSSION TITLE
Requiring the material of a precast concrete beam to be concrete

### DISCUSSION THREAD

**Discussion by RickBrice** [buildingSMART/IDS#389] [Q&A] [answered]
I'm trying to define an IDS for precast concrete bridge beams. I'm stumped at how to require the Category attribute of the IfcMaterial associated with IfcBeam to be concrete.

Actually, I'm not every sure how to intelligently ask this question.

Attached is an example IFC file and the IDS.

[PGSuper1.ifc.txt](https://github.com/user-attachments/files/17909558/PGSuper1.ifc.txt)
[Precast_Prestressed_Concrete_Beams.ids.txt](https://github.com/user-attachments/files/17909560/Precast_Prestressed_Concrete_Beams.ids.txt)

**Comment by Moult** [comment]
I tend to break down my material requirements into two parts.

First, I require that all materials must be of a valid category (the IDS material facet targets both material categories and names):

![2024-11-26-111700_840x216_scrot](https://github.com/user-attachments/assets/98e47dc3-1e96-4204-bfbe-de6130c2486b)

Then, I specify additional requirements on objects of different materials, such as saying "all concrete stuff must specify whether its insitu or precast":

![2024-11-26-111638_839x335_scrot](https://github.com/user-attachments/assets/3e51d33c-d52d-40a5-a87a-925fad87e97b)

**Reply by RickBrice** [reply]
That seems less definitive than saying all IfcBeam with Pset_ConcreteElementGeneral with property AssemblyPlace=FACTORY and ConstructionMethod=PRECAST must be associated with an IfcMaterial with Category attribute = 'concrete'

Other than comment sense, what prevents a beam with Pset_ConcreteElementGeneral properties from associating with an IfcMaterial.Category='steel'.

Probably silly questions - this is my first exploration of IDS

**Reply by Moult** [reply]
You can do what you're saying too, just flip the requirement:

 - Applies to: ConstructionMethod = PRECAST
 - Requirement: material = concrete

But I guess in my own scenarios I've found it more meaningful to start from the material: every material must be correct, and depending on the material, I require specific information (steel data for steel, concrete data for concrete, etc). I find it more intuitive to say to engineers "Hey, first tell me if you've got concrete, if you've got anything concrete, let me know if it's precast, and where it's assembled.

> what prevents a beam with Pset_ConcreteElementGeneral properties from associating with an IfcMaterial.Category='steel'.

Nothing in IFC prevents this. But if you write rules in IDS you can create this restriction.

In practice IDS is still very new and missing some (really basic?) stuff like negation, exclusions, or things like unioning groups of facets. I'd keep requirements as simple as possible, and still rely on digital engineers to do their job. Nothing prevents invalid data completely (I can call my Steel material Concrete if I want), so IDS is just another tool alongside visual inspection, model diffs, qualitative checks on scheduled output, etc.

**Reply by RickBrice** [reply]
That’s good information to think about. Do you know of any articles or blog posts that discuss the limits of IDS?

**Reply by atomczak** [reply]
I have an initial list, which is far from comprehensive (and some aspects are prone to change in the future):
✔️ element is of a certain class (entity)
✔️ element has a property
✔️ if a property value is within range
✔️ if a property value follows a pattern
✔️ element is required/optional/prohibited (cardinality)
✔️ property has right data type
✔️ classification is provided
✔️ material (name/category) associated with an element
✔️ containment (part of) (ex: furniture belongs to space)
✔️ IFC schema version
✔️ any combination of the above
❌ geometry exists
❌ geometry type (mesh, brep...)
❌ uniqueness (ex: can't have two walls with same 'mark')
❌ calculated values (ex: concrete totals)
❌ comparative values (ex: A bigger than B)
❌ connectivity (ex: pipeFitting is connected to pipeSegments)
❌ proximity (ex: door within 10m)
❌ who created it
❌ inheritance (IfcWall -> IfcBuiltElement)

**Comment by RickBrice** [comment]
Thanks @atomczak. Can you expand on "material (name/category) associated with an element"? That is exactly what I'm trying to accomplish but I can't seem to get there directly.

By guessing I've come up with a 3-specification solution:
1. Precast concrete beams must have a material with a name that corresponds to a pattern 
![image](https://github.com/user-attachments/assets/3992b86b-cda4-418a-aa1c-0831c8188221)

2. The model must have IfcMaterial with Category=concrete or steel (I'll need steel later for the reinforcements)
![image](https://github.com/user-attachments/assets/88328476-ba37-4792-8cc8-225b14c6020c)

3. IfcMaterial with Name=pattern, must have Category=concrete
![image](https://github.com/user-attachments/assets/8415281d-247b-42cd-aa25-c71d7820e303)

This approach seems rather indirect. Perhaps it's just the IDS tool I'm using?

I was hoping that I could do this with one specification. In my mind, I want to add "**MUST** have a material which **Category** is **concrete**" to the first specification shown above.

Here is my IDS file.
[Precast_Prestressed_Concrete_Beams.ids.txt](https://github.com/user-attachments/files/17923329/Precast_Prestressed_Concrete_Beams.ids.txt)

**Comment by andyward** [comment, marked as answer]
IDS's Material Facet's value is supposed to check both an IfcMaterial's **Name** and **Category**  ([See docs](https://github.com/buildingSMART/IDS/blob/development/Documentation/UserManual/material-facet.md#parameters)). So I think you should be able to express it as

**Applicable to**: _All IfcBeams that have a CastingMethod of 'PRECAST' and AssemblyMethod 'FACTORY'_
**Requirement**: _Should have any Material with value ='concrete'_

My updated IDS here
[Precast_Prestressed_Concrete_Beams.ids](https://github.com/user-attachments/files/17924612/Precast_Prestressed_Concrete_Beams.ids.txt)

Worth noting that this will also succeed for materials with a name of 'concrete' as well.
